src/report/bulk.rs

```diff
@@ -0,0 +1,202 @@
+//! A wire protocol for representing ekotrace log reports
+//! that are fragmented into multiple chunks due to sizing
+//! constraints
+//!
+use crate::compact_log::CompactLogItem;
+use crate::history::DynamicHistory;
+use crate::{ExtensionBytes, ReportError, TracerId};
+use core::mem::{align_of, size_of};
+use static_assertions::{assert_eq_align, const_assert_eq};
+
+/// Write reports all at once into a single byte slice.
+pub trait BulkReporter {
+    /// Conduct necessary background activities and write
+    /// the recorded reporting log to a collection backend,
+    /// including arbitrary extension bytes provided.
+    ///
+    /// Writes the Tracer's internal state according to the
+    /// log reporting schema.
+    ///
+    /// If the write was successful, returns the number of *bytes* written
+    fn report_with_extension(
+        &mut self,
+        destination: &mut [u8],
+        extension_metadata: ExtensionBytes<'_>,
+    ) -> Result<usize, ReportError>;
+
+    /// Conduct necessary background activities and write
+    /// the recorded reporting log to a collection backend.
+    ///
+    /// Writes the Tracer's internal state according to the
+    /// bulk log reporting format.
+    ///
+    /// If the write was successful, returns the number of *bytes* written
+    fn report(&mut self, destination: &mut [u8]) -> Result<usize, ReportError> {
+        self.report_with_extension(destination, ExtensionBytes(&[]))
+    }
+}
+impl<'log> BulkReporter for BulkReportSourceComponents<'log> {
+    fn report_with_extension(
+        &mut self,
+        destination: &mut [u8],
+        extension_metadata: ExtensionBytes<'_>,
+    ) -> Result<usize, ReportError> {
+        let n_log_bytes = self.log.len() * size_of::<CompactLogItem>();
+        let n_extension_bytes = extension_metadata.0.len();
+        if n_log_bytes > core::u32::MAX as usize {
+            // We don't currently support sending reports with logs that big.
+            // N.B. Point for future improvement - actually only send some of the log
+            // by doing some pre-segmentation and determining where we can cut it up,
+            // possibly injecting duplicate clocks as necessary.
+            return Err(ReportError::Encoding);
```

Comment:
This error value doesn't seem right

Reply:
The errors are intentionally sparse and suggestive rather than detailed in this library. 

Is there a particular thing you're hoping to capture? A different extant variant you'd prefer used, or are you asking for a new ReportError variant?

Reply:
If that's the intent it's fine; it was just a little surprising giving the specificity of the error below.

Reply:
That gap largely comes from the differences between errors bound for use in the C API versus errors that will likely only ever be exposed in the Rust API.